schema_bump_dry_run_scripts/genes/gene_bump_dry_run.py

```diff
@@ -0,0 +1,227 @@
+
+
+def fetch_private_datasets(base_url) -> Tuple[List[dict], Optional[str]]:
+    """
+    Fetches all private collections and parses the datasets from the response.
+    :param base_url:
+    :return:
+    """
+    auth_headers = get_headers(base_url)
+    for collection in fetch_private_collections(base_url, auth_headers):
+        collection_id = collection["collection_id"]
+        for ds in collection["datasets"]:
+            dataset_metadata = fetch_private_dataset(base_url, auth_headers, collection_id, ds["dataset_id"])
+            # only process uploaded datasets
+            if "processing_status" not in dataset_metadata or dataset_metadata["processing_status"] != "SUCCESS":
+                continue
+            yield dataset_metadata, collection.get("revision_of")
+
+
+def compare_genes(
+    dataset: Dict[str, Any], diff_map: Dict[str, str], deprecated_datasets: defaultdict(list)
+) -> Tuple[Dict, bool]:
+    """
+    Compare genes in a dataset with the provided diff map and update the deprecated_datasets dictionary.
+
+    :param dataset: The dataset to compare genes for.
+    :type dataset: dict
+    :param diff_map: The map of organisms and their deprecated genes.
+    :type diff_map: dict
+    :param deprecated_datasets: The dictionary to store deprecated datasets.
+    :type deprecated_datasets: dict
+
+    :return: A tuple containing the updated deprecated_datasets dictionary and a flag indicating if any deprecated
+    genes were found.
+    :rtype: tuple
+    """
+    dataset_id = dataset.get("dataset_id") or dataset.get("id")
+    collection_id = dataset["collection_id"]
+    dataset_genes_to_compare = set(get_genes(dataset))
+    organisms = [organism_info["ontology_term_id"] for organism_info in dataset["organism"]]
+    is_deprecated_genes_found = False
+    deprecated_genes_in_dataset = set()
+    for organism in organisms:
+        deprecated_genes_source = diff_map.get(organism, set())
+        intersection_genes = dataset_genes_to_compare.intersection(deprecated_genes_source)
+        if intersection_genes:
+            deprecated_genes_in_dataset.update(intersection_genes)
+            is_deprecated_genes_found = True
+
+    if is_deprecated_genes_found:
+        dataset_group_key = (*sorted(deprecated_genes_in_dataset), len(dataset_genes_to_compare))
+        # dataset_group_key is a unique key used to identify a group of datasets that have the same set of deprecated
+        # genes and the same number of genes. This is used to group datasets together in the report.
+        if collection_id not in deprecated_datasets:
+            deprecated_datasets[collection_id] = {"dataset_groups": {}}
+        collection = deprecated_datasets[collection_id]
+        if dataset_group_key not in collection["dataset_groups"]:
+            collection["dataset_groups"][dataset_group_key] = {
+                "datasets": [],
+                "num_datasets": 0,
+                "deprecated_genes": deprecated_genes_in_dataset,
```

Comment:
nit: do we need these if we can derive them from the dataset_group_key?

Reply:
these == 'deprecated genes' and 'num_genes'

Reply:
> nit: do we need these if we can derive them from the dataset_group_key?
yes we could, but for readability I treated them as different things. If we did use the dataset_group_key we would need to parse the tuple as `dataset_group_key[:-1]` which makes less sense than using `deprecated_genes_in_dataset` which has already be calculated and stored. This would al require moving more processing logic into the parent function which would be more code duplication.

Reply:
> these == 'deprecated genes' and 'num_genes'
`num_genes` is the number of genes in a dataset, not just deprecated. This was a specific request from to track these values.